pkg/scheduler/framework/preemption/preemption.go

```diff
@@ -0,0 +1,574 @@
+		n = int32(len(cl.items))
+	}
+	return n
+}
+
+// get returns the internal candidate array. This function is NOT atomic and
+// assumes that all add() operations have been completed.
+func (cl *candidateList) get() []Candidate {
+	return cl.items[:cl.size()]
+}
+
+// Inteface is expected to be implemented by different preemption plugins as all those member
+// methods might have different behavior compared with the default preemption.
+type Inteface interface {
+	// GetOffsetAndNumCandidates chooses a random offset and calculates the number of candidates that should be
+	// shortlisted for dry running preemption.
+	GetOffsetAndNumCandidates(nodes int32) (int32, int32)
+	// CandidatesToVictimsMap builds a map from the target node to a list of to-be-preempted Pods and the number of PDB violation.
+	CandidatesToVictimsMap(candidates []Candidate) map[string]*extenderv1.Victims
+	// PodEligibleToPreemptOthers determines whether this pod should be considered
+	// for preempting other pods or not.
+	PodEligibleToPreemptOthers(pod *v1.Pod, nodeInfos framework.NodeInfoLister, nominatedNodeStatus *framework.Status) bool
+	// SelectVictimsOnNode finds minimum set of pods on the given node that should be preempted in order to make enough room
+	// for "pod" to be scheduled.
+	SelectVictimsOnNode(ctx context.Context, state *framework.CycleState,
+		pod *v1.Pod, nodeInfo *framework.NodeInfo, pdbs []*policy.PodDisruptionBudget) ([]*v1.Pod, int, *framework.Status)
+}
+
+type Evaluator struct {
+	PluginNameName string
+	Context        context.Context
+	Handler        framework.Handle
+	PodLister      corelisters.PodLister
+	PdbLister      policylisters.PodDisruptionBudgetLister
+	State          *framework.CycleState
+	Inteface
+}
+
+func (ev *Evaluator) Preempt(pod *v1.Pod, m framework.NodeToStatusMap) (*framework.PostFilterResult, *framework.Status) {
+	cs := ev.Handler.ClientSet()
+	nodeLister := ev.Handler.SnapshotSharedLister().NodeInfos()
+
+	// 0) Fetch the latest version of <pod>.
+	// It's safe to directly fetch pod here. Because the informer cache has already been
+	// initialized when creating the Scheduler obj, i.e., factory.go#MakeDefaultErrorFunc().
+	// However, tests may need to manually initialize the shared pod informer.
+	podNamespace, podName := pod.Namespace, pod.Name
+	pod, err := ev.PodLister.Pods(pod.Namespace).Get(pod.Name)
+	if err != nil {
+		klog.ErrorS(err, "getting the updated preemptor pod object", "pod", klog.KRef(podNamespace, podName))
+		return nil, framework.AsStatus(err)
+	}
+
+	// 1) Ensure the preemptor is eligible to preempt other pods.
+	if !ev.PodEligibleToPreemptOthers(pod, nodeLister, m[pod.Status.NominatedNodeName]) {
+		klog.V(5).InfoS("Pod is not eligible for more preemption", "pod", klog.KObj(pod))
+		return nil, framework.NewStatus(framework.Unschedulable)
+	}
+
+	// 2) Find all preemption candidates.
+	candidates, nodeToStatusMap, status := ev.findCandidates(ev.Context, ev.State, pod, m)
```

Comment:
ev.State can be derived from ev, so don't need to have `state` in the signature of findCandidates().

Reply:
Done